ivre/analyser.py

```diff
@@ -0,0 +1,667 @@
+                                               "count": "$count"}},
+                        "count": {"$sum": "$count"}}},
+            {'$sort': {"count": -1}},
+        ]
+        result = {}
+        for rec in db.nmap.db[db.nmap.colname_hosts].aggregate(pipeline,
+                                                               cursor={}):
+            result[rec['_id']] = {srvprod['name']:
+                                  float(srvprod['count']) / host_count
+                                  for srvprod in rec['srvprods']}
+            result[rec['_id']][u'UNAVAILABLE'] = 1 - \
+                float(rec['count']) / host_count
+        return result
+
+    def _explain_components(self, variance, pca_indexes,
+                            n_comp=10, dimension=2):
+        """
+        Plot n_comp most relevant components for the first dimensions
+        """
+        assert dimension <= 3
+        # Prevent from index error in case PCA went on fewer dimension
+        n_comp = min(n_comp, len(pca_indexes[0]), len(pca_indexes[1]))
+        fig = plt.figure()
+        AXIS = ['X', 'Y', 'Z']
+        for dim in xrange(dimension):
+            plot = fig.add_subplot(dimension, 1, dim+1)
+            values = [pca_indexes[dim][i][1] for i in xrange(n_comp)]
+            ind = np.arange(n_comp)
+            width = 0.35
+            plot.bar(ind, values, width, color='cyan')
+            marks = [self._matrix_col_mean[pca_indexes[dim][i][0]]
+                     for i in xrange(n_comp)]
+            plot.set_xticks(ind + width / 2.0)
+            plot.set_xticklabels(marks)
+            plot.set_xlim(-width, len(ind) + width / 2.0)
+            margin = 0.1 * max(abs(i) for i in values)
+            plot.set_ylim(min(values) - margin, max(values) + margin)
+            plot.legend(['Variance : %f' % variance[dim]])
+            plot.set_title('%s axis decomposition' % AXIS[dim])
+
+    @staticmethod
+    def _pca(hosts, dim=2):
+        """
+        Principal component analysis
+        Reduce the numpy-matrix hosts to a dim-dimensional vector space
+        """
+        pca = RandomizedPCA(n_components=dim)
+        pca.fit(hosts)
+        # Return most discriminating values by axis.
+        pca_indexes = []
+        for ind in xrange(dim):
+            vect = pca.components_[ind]
+            pca_indexes.append([(idx, vect[idx])
+                                for idx in (-abs(vect)).argsort()])
+        return (pca, pca_indexes)
+
+    @staticmethod
+    def _search_matrix_extremum(hosts):
+        """
+        Return two lists containing minimum and maximum values
+        by dimensions for the numpy matrix hosts
+        """
+        min_size = []
+        max_size = []
+        dim = hosts.shape[1]
+        for idx in xrange(dim):
+            min_size.append(round(hosts[:, idx].min() - 1))
+            max_size.append(round(hosts[:, idx].max() + 1))
+        return (min_size, max_size)
+
+    def plot(self, dim=2):
```

Comment:
If the only valid values for `dim` are 2 and 3, use a boolean like `plot3d` with default value `False`.